src/libmongoc/tests/test-mongoc-counters.c

```diff
@@ -481,6 +471,189 @@ test_counters_streams_timeout (void)
+      mongoc_client_set_error_api (client, MONGOC_ERROR_API_VERSION_2);
+   }
+
+   mongoc_counter_auth_success_reset ();
+   mongoc_counter_auth_failure_reset ();
+   mongoc_counter_op_egress_query_reset ();
+   mongoc_counter_op_egress_msg_reset ();
+
+   ASSERT_OR_PRINT (
+      mongoc_client_get_server_status (client, NULL, NULL, &error), error);
+
+   const int32_t auth_success = mongoc_counter_auth_success_count ();
+   const int32_t auth_failure = mongoc_counter_auth_failure_count ();
+   const int32_t sent_queries = mongoc_counter_op_egress_query_count ();
+   const int32_t sent_msgs = mongoc_counter_op_egress_msg_count ();
+
+   // Ensure we are not testing more than we intend.
+   ASSERT_WITH_MSG (auth_success == 1 && auth_failure == 0,
+                    "expected exactly one authentication attempt to succeed, "
+                    "but observed %" PRId32 " successes and %" PRId32
+                    " failures",
+                    auth_success,
+                    auth_failure);
+
+   // MongoDB Handshake Spec: Since MongoDB server 4.4, the initial handshake
+   // supports a new argument, `speculativeAuthenticate`, provided as a BSON
+   // document. Clients specifying this argument to hello or legacy hello will
+   // speculatively include the first command of an authentication handshake.
+   const bool has_speculative_auth = test_framework_get_server_version () >=
+                                     test_framework_str_to_version ("4.4.0");
+
+   // The number of expected OP_QUERY requests depends on pooling and the
+   // presence of the RTT monitor thread.
+   if (pooled) {
+      // RTT monitoring is also a 4.4+ feature alongside speculative
+      // authentication and affects the number of OP_QUERY requests.
+      if (has_speculative_auth) {
+         // OP_QUERY requests consists of:
+         //  - initial connection handshake by server monitor thread (x3)
+         //  - initial connection handshake by RTT monitor thread (x3)
+         //  - polling hello by RTT monitor thread (x3)
+         //  - initial connection handshake by new cluster node
+         ASSERT_WITH_MSG (
+            sent_queries == 10,
+            "expected exactly ten OP_QUERY requests, but observed %" PRId32
+            " requests",
+            sent_queries);
+      } else {
+         // OP_QUERY requests consists of:
+         //  - initial connection handshake by server monitor (x3)
+         //  - initial connection handshake by new cluster node
+         ASSERT_WITH_MSG (
+            sent_queries == 4,
+            "expected exactly four OP_QUERY requests, but observed %" PRId32
+            " requests",
+            sent_queries);
+      }
+   } else {
+      // OP_QUERY requests consists of:
+      //  - initial connection handshake (x3)
+      ASSERT_WITH_MSG (
+         sent_queries == 3,
+         "expected exactly three OP_QUERY requests, but observed %" PRId32
+         " requests",
+         sent_queries);
```

Comment:
This assertion fails on my machine with `expected exactly three OP_QUERY requests, but observed 6 requests`.

This led to bug report in the C driver: CDRIVER-4622
OP_QUERY egress appears to be overcounted when the hostname resolves to multiple DNS records. On my machine, "localhost" resolves to both IPv4 `127.0.0.1` and IPv6 `::1`.

I suggest adding a function to count the DNS records of "localhost" (modified from `test_framework_skip_if_no_dual_ip_hostname`):
```c
// get_dns_record_count returns the number of DNS records for "localhost".
// "localhost" may resolve to both the IPv4 `127.0.0.1` and IPv6 `::1`.
int
get_dns_record_count (void)
{
   struct addrinfo hints = {0}, *res = NULL, *iter;
   int res_count = 0;
   char *host = "localhost";

   hints.ai_family = AF_UNSPEC;
   hints.ai_socktype = SOCK_STREAM;
   hints.ai_flags = 0;
   hints.ai_protocol = 0;

   BSON_ASSERT (getaddrinfo (host, "27017", &hints, &res) != -1);

   iter = res;

   while (iter) {
      res_count++;
      iter = iter->ai_next;
   }

   freeaddrinfo (res);

   return res_count;
}
```
And conditionally skipping this test:
```c
   if (get_dns_record_count () > 1) {
      MONGOC_DEBUG (
         "Skipping test due to overcounting bug described in CDRIVER-4622");
      return;
   }
```